gitpod-io/gitpod
# ws-mananger: refactor volume snapshot watcher

## Description

Refactor volume snapshot watcher.

Using the Go channel to notify different event workers that the volume snapshot is ready seems risky.
We observe that at gen67 cluster, some workspaces with PVC can't be terminated. The symptom is the workspace pod starts the disposal process, and the volume snapshot is ready to use. But somehow, the event does not receive by the receiver _or_ does not send to the sender.

We rewrite the watcher with the volume snapshot client watch + retry watcher to make sure the watcher happens in each event worker. To minimize the chance that the event does not receive by the receiver _or_ does not send to the sender.
Moreover, we fall back to exponential retry to get the volume snapshot object if we can't create the retry watcher.

## Related Issue(s)

Fixes #13280

## How to test

Create an ephemeral cluster, and run loadgen test with 100 regular workspaces + 100 regular workspaces with PVC.
Ensure no workspace pod is terminating for a long time.

## Release Notes

```release-note
None
```

## Documentation

None

## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`
